File: winbolo/src/bolo/mines.c
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */


/*********************************************************
*Name:          Mines
*Filename:      mines.c 
*Author:        John Morrison
*Creation Date: 29/1/99
*Last Modified: 2/11/99
*Purpose:
*  Handles what mines are visible to the user
*********************************************************/

#include "global.h"
#include "bolo_map.h" 
#include "screen.h"
#include "mines.h"

/* Prototypes */

/*********************************************************
*NAME:          minesCreate
*AUTHOR:        John Morrison
*CREATION DATE: 29/1/98
*LAST MODIFIED: 29/1/98
*PURPOSE:
*  Creates and initilises the mines structure.
*
*ARGUMENTS:
*  value - Pointer to the map file
*********************************************************/
void minesCreate(mines *visMines, bool allowHiddenMines) {
  int count1; /* Looping Variables */
  int count2;

  New(*visMines);
  for (count1=0;count1<MINES_ARRAY_SIZE;count1++) {
    for (count2=0;count2<MINES_ARRAY_SIZE;count2++) {
      (*visMines)->pos[count1][count2] = FALSE;
    }
  }
  (*visMines)->minesHiddenMines = allowHiddenMines;
}

/*********************************************************
*NAME:          mapDestroy
*AUTHOR:        John Morrison
*CREATION DATE: 29/10/98
*LAST MODIFIED: 29/10/98
*PURPOSE:
*  Destroys the mines data structure. Also frees memory.
*
*ARGUMENTS:
*
*********************************************************/
void minesDestroy(mines *visMines) {
  Dispose(*visMines);
}

/*********************************************************
*NAME:          minesGetAllowHiddenMines
*AUTHOR:        John Morrison
*CREATION DATE: 29/1/99
*LAST MODIFIED: 29/1/99
*PURPOSE:
*  Returns whether hidden mines are allowed or not
*
*ARGUMENTS:
*
*********************************************************/
bool minesGetAllowHiddenMines(mines *visMines) {
  return (*visMines)->minesHiddenMines;
}

/*********************************************************
*NAME:          minesAddItem
*AUTHOR:        John Morrison
*CREATION DATE: 29/1/99
*LAST MODIFIED:  3/2/99
*PURPOSE:
* Adds a mine to the structure. Returns whether a mine
* already existed at that position.
*
*ARGUMENTS:
*  xValue   - X Map Coordinate
*  yValue   - Y Map Coordinate
*********************************************************/
bool minesAddItem(mines *visMines, BYTE xValue, BYTE yValue) {
  bool returnValue = FALSE; /* Value to return */

  returnValue = (*visMines)->pos[xValue][yValue];
  (*visMines)->pos[xValue][yValue] = TRUE;
  return returnValue;
}

/*********************************************************
*NAME:          minesRemoveItem
*AUTHOR:        John Morrison
*CREATION DATE: 8/1/99
*LAST MODIFIED: 8/2/99
*PURPOSE:
* Removes a mine to the structure.
*
*ARGUMENTS:
*  xValue   - X Map Coordinate
*  yValue   - Y Map Coordinate
*********************************************************/
void minesRemoveItem(mines *visMines, BYTE xValue, BYTE yValue) {
  (*visMines)->pos[xValue][yValue] = FALSE;
}


/*********************************************************
*NAME:          minesExistPos
*AUTHOR:        John Morrison
*CREATION DATE: 29/1/99
*LAST MODIFIED: 29/1/99
*PURPOSE:
* Returns whether a mine can be seen at that position.
* Only called if a mine does exist at the map square
* but checking here to see if the player knows about it
*
*ARGUMENTS:
*  xValue   - X Map Coordinate
*  yValue   - Y Map Coordinate
*********************************************************/
bool minesExistPos(mines *visMines, BYTE xValue, BYTE yValue) {
  bool returnValue = TRUE; /* Value to return */
  
  if (xValue <= MAP_MINE_EDGE_LEFT || xValue >= MAP_MINE_EDGE_RIGHT || yValue <= MAP_MINE_EDGE_TOP || yValue >= MAP_MINE_EDGE_BOTTOM) {
    returnValue = TRUE;
  } else if ((*visMines)->minesHiddenMines == TRUE) {
    returnValue = (*visMines)->pos[xValue][yValue];
  } else {
    returnValue = screenMapIsMine(xValue, yValue);
  }
  return returnValue;
}

File: jbolo/src/com/jbolo/bolo/Mines.java
/*
 * $Id$
 *
 * Copyright (c) 1998-2008 John Morrison.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */



package com.jbolo.bolo;

import com.jbolo.bolo.map.BoloMap;

/*
 * Mines.java
 *
 * Created on October 7, 2003, 8:08 AM
 */

/** Object stores whether our tank knows whether mines are on the map
 * @author John Morrison
 */
public class Mines {
    
    /** Mine positions as known by our tank */    
    private boolean[][] pos;
    /** Does this game have hidden mines */    
    private boolean allowHiddenMines;
    
    /** GameWorld that owns this */
    private GameWorld gw;
    
    /** Creates a new instance of Mines
     *
     * @param gw GameWorld that owns this
     * @param allowHiddenMines Does this game allow hidden mines
     */
    public Mines(GameWorld gw, boolean allowHiddenMines) {
        int count;
        int count2;
        
        this.allowHiddenMines = allowHiddenMines;
	this.gw = gw;
        pos = new boolean[256][256];
        for (count=0;count<256;count++) {
            for (count2=0;count2<256;count2++) {
                pos[count][count2] = false;
            }
        }
    }
    
    /** Returns if this game allows hidden mines
     * @return boolean
     */    
    public boolean getAllowHiddenMines() {
        return allowHiddenMines;
    }

    /** Sets if this game allows hidden mines
     * @param value Value to set
     */    
    public void setAllowHiddenMines(boolean value) {
        allowHiddenMines = value;
    }
    
    
    /** Adds a mine to this tanks known mines
     * @param x X map co-ordinate to add a mine at
     * @param y Y map co-ordinate to add a mine at
     * @return If a mine existed in this position already
     */    
    public boolean add(int x, int y) {
        boolean returnValue; // Value to return
        if (x <= BoloMap.MAP_MINE_EDGE_LEFT || x >= BoloMap.MAP_MINE_EDGE_RIGHT || y <= BoloMap.MAP_MINE_EDGE_TOP || y >= BoloMap.MAP_MINE_EDGE_BOTTOM) {
            return true;
        }
        returnValue = pos[x][y];
        pos[x][y] = true;
        return returnValue;
    }
    
    /** Removes a mine from the tanks known mines
     * @param x X map co-ordinate to remove mine from
     * @param y Y map co-ordinate to remove mine from
     */    
    public void remove(int x, int y) {
       pos[x][y] = false;   
    }
    
    /** existPos checks if a mine is known to the tank at map position x,y
     * @param x X map co-ordinate to check
     * @param y Y map co-ordinate to check
     * @return boolean
     */    
    public boolean existPos(int x, int y) {
        if (x <= BoloMap.MAP_MINE_EDGE_LEFT || x >= BoloMap.MAP_MINE_EDGE_RIGHT || y <= BoloMap.MAP_MINE_EDGE_TOP || y >= BoloMap.MAP_MINE_EDGE_BOTTOM) {
            return true;
        } else if (allowHiddenMines == true) {
            return pos[x][y];
        }
    
	return gw.getBoloMap().isMine(x, y);

    }
}
